llvm/include/llvm/IR/RepoDefinition.h

```diff
@@ -107,56 +102,110 @@ struct GODigestState {
+  /// GO's hash is cached.
+  std::tuple<DigestType, bool> getGOHash(const GlobalObject *GO) const {
+    const auto Iter = GOHashCache.find_as(GO);
+    return (Iter != GOHashCache.end()) ? std::make_tuple(Iter->second, true)
+                                       : std::make_tuple(NullDigest, false);
   }
-  // Update G's dependencies.
-  GOInfo &GInfo = GOI[G];
-  GInfo.InitialDigest = std::move(Result.InitialDigest);
-  GInfo.Dependencies = std::move(Result.Dependencies);
-}
-
-// Create a global object information map and calculate the number of hashed
-// functions and variable  inside of the Module M.
-/// \param M Called module.
-/// \returns a pair containing a global object information map and a structure
-/// holding the number of hashed global variables and functions.
-std::pair<GOInfoMap, GONumber> calculateGONumAndGOIMap(Module &M);
+
+  /// Construct the global object information map (GOIMap) and calculate the
+  /// number of hashed functions and variable  inside of the Module M.
+  void calculateGONumAndGOIMap(const Module &M);
+
+  /// Compute the hash value for the given global object GO.
+  /// \param GO The global object.
+  /// \param Final true if using the digest of GO's dependencies which is
+  ///        stored in the RepoDefinition metadata.
+  /// \return The global object's digest value.
+  DigestType calculateDigest(const GlobalObject *GO, bool Final);
+
+  /// Calculate the initial hash value, dependencies and contributions for the
+  /// global object 'G' and store these informaion in the GOImap.
+  /// \param G Calculated global object.
+  template <typename GlobalType> void calculateGOInfo(const GlobalType *G);
+  void calculateGOInfo(const GlobalObject *GO);
+
+  /// Accumulate the GO's digest and get its GODigestState.
+  GODigestState accumulateGODigest(const GlobalObject *GO, MD5 &GOHash,
+                                   bool Final);
+
+  /// Loop through the contributions and add the initial digest of each global
+  /// object inside of the contributions to GOHash.
+  void updateDigestUseContributions(MD5 &GOHash, const GOVec &Contributions);
+
+  /// Loop through the dependences and add the final digest of each global
+  /// object inside of the dependences to GOHash.
+  std::tuple<size_t, DigestType>
+  updateDigestUseDependencies(const GlobalObject *GO, MD5 &GOHash,
+                              unsigned GOVisitedIndex,
+                              const GOVec &Dependencies, bool Final);
+
+  /// Update the digest of an invidual GO incorporating the hashes of all its
+  /// dependencies and contributions.
+  /// \param GO The global object whose digest is to be computed.
+  /// \param Final true if using the digest of GO's dependencies which is
+  ///        stored in the RepoDefinition metadata.
+  /// \returns a tuple containing the state information which includes
+  ///          1) a numerical identifier which will be used if GO loops back to
```

Comment:
> a numerical identifier which will be used if GO loops back to itself in future

I think it is better to describe what this number represents rather than its usage. I think that it is the lowest visited index of any dependency visited by this call (or max if there are no dependencies).

That (hopefully) make it clear how it could be used for detecting loops. If you find a lower index than GO in its dependencies then you've gone back on yourself (for this particular traversal).

Reply:
Changed to: `the depth of the path that we've traversed (or max if it reaches the end of the path)`.